File: clients/Go/model/auto_attack.go
package model

import "io"
import . "aicup2020/stream"

type AutoAttack struct {
    PathfindRange int32
    ValidTargets []EntityType
}
func NewAutoAttack(pathfindRange int32, validTargets []EntityType) AutoAttack {
    return AutoAttack {
        PathfindRange: pathfindRange,
        ValidTargets: validTargets,
    }
}
func ReadAutoAttack(reader io.Reader) AutoAttack {
    result := AutoAttack {}
    result.PathfindRange = ReadInt32(reader)
    result.ValidTargets = make([]EntityType, ReadInt32(reader))
    for i := range result.ValidTargets {
        result.ValidTargets[i] = ReadEntityType(reader)
    }
    return result
}
func (value AutoAttack) Write(writer io.Writer) {
    WriteInt32(writer, value.PathfindRange)
    WriteInt32(writer, int32(len(value.ValidTargets)))
    for _, ValidTargetsElement := range value.ValidTargets {
        WriteInt32(writer, int32(ValidTargetsElement))
    }
}

File: clients/FSharp/Model/AutoAttack.fs
#nowarn "0058"
namespace Aicup2020.Model
type AutoAttack = {
    PathfindRange: int;
    ValidTargets: EntityType[];
    } with
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write this.PathfindRange
        writer.Write this.ValidTargets.Length
        this.ValidTargets |> Array.iter (fun value ->
            writer.Write (int value)
        )
    static member readFrom(reader: System.IO.BinaryReader) = {
        PathfindRange = reader.ReadInt32()
        ValidTargets = [|for _ in 1 .. reader.ReadInt32() do
            yield reader.ReadInt32() |> enum
        |]
    }
